bindings/java/java/src/main/java/org/ethereum/evmc/EvmcVm.java

```diff
@@ -0,0 +1,113 @@
+
+  static {
+    try {
+      System.load(System.getProperty("user.dir") + "/../c/build/evmc.so");
+    } catch (UnsatisfiedLinkError e) {
+      System.err.println("Native code library failed to load.\n" + e);
+      System.exit(1);
+    }
+  }
+
+  public static native int test(ByteBuffer buff);
+
+  public static native int init(String filename);
+
+  /**
+   * EVMC ABI version implemented by the VM instance.
+   *
+   * <p>Can be used to detect ABI incompatibilities. The EVMC ABI version represented by this file
+   * is in ::EVMC_ABI_VERSION.
+   */
+  public static native int abi_version();
+
+  /**
+   * The name of the EVMC VM implementation.
+   *
+   * <p>It MUST be a NULL-terminated not empty string. The content MUST be UTF-8 encoded (this
+   * implies ASCII encoding is also allowed).
+   */
+  public static native String name();
+
+  /**
+   * The version of the EVMC VM implementation, e.g. "1.2.3b4".
+   *
+   * <p>It MUST be a NULL-terminated not empty string. The content MUST be UTF-8 encoded (this
+   * implies ASCII encoding is also allowed).
+   */
+  public static native String version();
+
+  /**
+   * Function to destroy the VM instance.
+   *
+   * <p>This is a mandatory method and MUST NOT be set to NULL.
+   */
+  public static native void destroy();
+
+  /**
+   * Function to execute a code by the VM instance.
+   *
+   * <p>This is a mandatory method and MUST NOT be set to NULL.
+   */
+  public static native ByteBuffer execute(
+      int context_index, int rev, ByteBuffer msg, ByteBuffer code, int size);
+  /**
+   * Function is a wrapper around native execute.
+   *
+   * <p>This allows the context to managed in one method
+   */
+  public static synchronized ByteBuffer execute(
+      HostContext context, int rev, ByteBuffer msg, ByteBuffer code, int size) {
+    int context_index = addContext(context);
+    ByteBuffer result = execute(context_index, rev, msg, code, size);
+    removeContext(context_index);
+    return result;
+  }
+
+  /**
+   * A method returning capabilities supported by the VM instance.
+   *
+   * <p>The value returned MAY change when different options are set via the set_option() method.
+   *
+   * <p>A Client SHOULD only rely on the value returned if it has queried it after it has called the
+   * set_option().
+   *
+   * <p>This is a mandatory method and MUST NOT be set to NULL.
+   */
+  public static native ByteBuffer get_capabilities();
+
+  /**
+   * Function that modifies VM's options.
+   *
+   * <p>If the VM does not support this feature the pointer can be NULL.
+   */
+  public static native int set_option(String name, String value);
+
+  private static List<HostContext> contextList =
+      Collections.synchronizedList(new ArrayList<HostContext>());
+
+  public static HostContext getContext(int index) {
+    return contextList.get(index);
+  }
+
+  public static int addContext(HostContext context) {
+    contextList.add(context);
+    return contextList.size() - 1;
```

Comment:
This would be potential place for a race condition if `addContext` would be called from different threads.

Reply:
thanks.  just added the synchronized keyword